preprocess.py

```diff
@@ -3,91 +3,220 @@
+        return [dataset]
+
+    # Divide the big corpus into shards, and build dataset seperately.
+    ret_list = []
+    src_iter = ShardedCorpusIterator(
+            src_corpus, opt.src_seq_length_trunc,
+            "src", opt.max_shard_size)
+    tgt_iter = ShardedCorpusIterator(
+            tgt_corpus, opt.tgt_seq_length_trunc,
+            "tgt", opt.max_shard_size, assoc_iter=src_iter)
+
+    index = 0
+    while not src_iter.hit_end():
+        index += 1
+        dataset = onmt.IO.ONMTDataset(
+                src_iter, tgt_iter, fields,
+                opt.src_seq_length, opt.tgt_seq_length,
+                dynamic_dict=opt.dynamic_dict)
+        ret_list.append(dataset)
+
+        # We save fields in vocab.pt seperately, so make it empty.
+        dataset.fields = []
+        part_name = "{:s}.{:s}.{:d}.pt".format(
+                    opt.save_data, data_type, index)
+        torch.save(dataset, open(part_name, 'wb'))
+
+    return ret_list
 
 
 def main():
-    print('Preparing training ...')
-    with codecs.open(opt.train_src, "r", "utf-8") as src_file:
-        src_line = src_file.readline().strip().split()
-        _, _, n_src_features = onmt.IO.extract_features(src_line)
-    with codecs.open(opt.train_tgt, "r", "utf-8") as tgt_file:
-        tgt_line = tgt_file.readline().strip().split()
-        _, _, n_tgt_features = onmt.IO.extract_features(tgt_line)
+    opt = parse_args()
 
+    print("Preparing training...")
+    n_src_features = get_num_features(opt.train_src)
+    n_tgt_features = get_num_features(opt.train_tgt)
     fields = onmt.IO.get_fields(n_src_features, n_tgt_features)
-    print("Building Training...")
-    train = onmt.IO.ONMTDataset(
-        opt.train_src, opt.train_tgt, fields,
-        opt.src_seq_length, opt.tgt_seq_length,
-        src_seq_length_trunc=opt.src_seq_length_trunc,
-        tgt_seq_length_trunc=opt.tgt_seq_length_trunc,
-        dynamic_dict=opt.dynamic_dict)
-    print("Building Vocab...")
-    onmt.IO.build_vocab(train, opt)
-
-    print("Building Valid...")
-    valid = onmt.IO.ONMTDataset(
-        opt.valid_src, opt.valid_tgt, fields,
-        opt.src_seq_length, opt.tgt_seq_length,
-        src_seq_length_trunc=opt.src_seq_length_trunc,
-        tgt_seq_length_trunc=opt.tgt_seq_length_trunc,
-        dynamic_dict=opt.dynamic_dict)
-    print("Saving train/valid/fields")
 
-    # Can't save fields, so remove/reconstruct at training time.
-    torch.save(onmt.IO.save_vocab(fields),
-               open(opt.save_data + '.vocab.pt', 'wb'))
-    train.fields = []
-    valid.fields = []
-    torch.save(train, open(opt.save_data + '.train.pt', 'wb'))
-    torch.save(valid, open(opt.save_data + '.valid.pt', 'wb'))
+    print("Building & saving training data...")
+    train_datasets = build_save_dataset('train', fields, opt)
+
+    print("Building & saving validation data...")
+    build_save_dataset('valid', fields, opt)
+
+    print("Building & saving vocabulary...")
```

Comment:
99% sure you have to create the vocab before creating valid. The vocab should not know about the valid dataset.